include/psa/crypto.h

```diff
@@ -3601,6 +3643,100 @@ psa_status_t psa_key_derivation_output_key(
     psa_key_derivation_operation_t *operation,
     mbedtls_svc_key_id_t *key);
 
+/** Compare output data from a key derivation operation to an expected value.
+ *
+ * This function calculates output bytes from a key derivation algorithm and
+ * compares those bytes to an expected value.
+ * If you view the key derivation's output as a stream of bytes, this
+ * function destructively reads the requested number of bytes from the
+ * stream before comparing them.
+ * The operation's capacity decreases by the number of bytes read.
+ *
+ * If this function returns an error status other than
+ * #PSA_ERROR_INSUFFICIENT_DATA, the operation enters an error
+ * state and must be aborted by calling psa_key_derivation_abort().
+ *
+ * \param[in,out] operation The key derivation operation object to read from.
+ * \param[in] expected_output Buffer where the output will be written.
+ * \param output_length     Length ot the expected output; this is also the
+ *                          number of bytes that will be read.
+ *
+ * \retval #PSA_SUCCESS
+ * \retval #PSA_ERROR_INVALID_SIGNATURE
+ *         The output was read successfully, but if differs from the expected
+ *         output.
+ * \retval #PSA_ERROR_INSUFFICIENT_DATA
+ *                          The operation's capacity was less than
+ *                          \p output_length bytes. Note that in this case,
+ *                          the operation's capacity is set to 0, thus
+ *                          subsequent calls to this function will not
+ *                          succeed, even with a smaller output buffer.
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The operation state is not valid (it must be active and completed
+ *         all required input steps).
+ * \retval #PSA_ERROR_INSUFFICIENT_MEMORY
+ * \retval #PSA_ERROR_COMMUNICATION_FAILURE
+ * \retval #PSA_ERROR_HARDWARE_FAILURE
+ * \retval #PSA_ERROR_CORRUPTION_DETECTED
+ * \retval #PSA_ERROR_STORAGE_FAILURE
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The library has not been previously initialized by psa_crypto_init().
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_key_derivation_verify_output_bytes(
+    psa_key_derivation_operation_t *operation,
+    const uint8_t *output,
+    size_t output_length);
+
+/** Compare output data from a key derivation operation to an expected value.
+ *
+ * This function calculates output bytes from a key derivation algorithm and
+ * compares those bytes to an expected value, provided as key of type
+ * #PSA_KEY_TYPE_RAW_DATA.
```

Comment:
I'm not sure what distinction you're doing between `PASSWORD_HASH` and `DERIVATION_OUTPUT`, could you clarify?

Reply:
By “password hash”, I mean the output of a key stretching KDF, and by “key derivation output”, the output of a non-stretching KDF. I make a distinction because it's feasible to recover the password from its hash if the password isn't very strong, whereas it's infeasible to recover the high-entropy secret input to a KDF from its output.

Reply:
> [Do we need a new key type (or even two) other than RAW_DATA for the key input of verify_output_key() and how should we call it?](https://github.com/ARMmbed/mbedtls/pull/4377#issuecomment-830021213)

We at least need a type for password hashes, which is suitable as the output of `psa_key_derivation_output_key` on a key stretching function and as the input of `psa_key_derivation_verify_key`. Internally, I this key might even store the parameters used to generate the hash (salt, cost) as well as the raw output. I suggest the name `PSA_KEY_TYPE_PASSWORD_HASH`.

By analogy, it would make sense to have a dedicated type `PSA_KEY_TYPE_DERIVATION_OUTPUT` for the non-stretching case. However, unlike the strecthing case, I don't think it's a common scenario.

Generally, avoid `PSA_KEY_TYPE_RAW_DATA` for anything that has a cryptographic “meaning”: anything that has a specific structure (e.g. some asymmetric keys), randomness or uniqueness requirements (keys, nonces, …), etc. `PSA_KEY_TYPE_RAW_DATA` is ok for a label that has no generic expectation of secrecy or uniqueness.

Reply:
I'm not entirely sure we need that distinction here. Imagine a hash `h` is produced using a password hashing scheme `A`, so we know `h = A(p)` for some `p` in a relatively small set `S` (which we assume the attack can enumerate easily). Now assume an attacker can somehow trick us into using a different algorithm B for verification, that was not designed to handle passwords and is much faster for the attacker to evaluate. There's no reason there would exist a set `S'` that the attacker could easily enumerate and that would necessarily contain a `p'` such that `h = B(p')`.

So, really I think even if we wanted to support verification in the non-stretching case, I'm not sure a different key type would be necessary.

Anyway, for now I've introduced a single key type called `PASSWORD_HASH` as this is the thing that's used in real life, and out of precaution I've documented it as being usable only with password hashing algorithms. If in the future we find out people are doing that with non-stretching KDFs too, at this point we can always decide to either (1) keep only the existing type and lift the documented restriction - the name would just reflect the most common use case, or (2) keep the restriction on the existing type and add a second one.

So, I think the way it's done in the current version of this PR is safe and does not block further extension to the non-stretching case, should it become desirable.